Description: Install signed images if UEFI Secure Boot is enabled
Author: Colin Watson <cjwatson@ubuntu.com>
Forwarded: no
Last-Update: 2012-12-05

Index: b/util/grub-install.in
===================================================================
--- a/util/grub-install.in
+++ b/util/grub-install.in
@@ -60,6 +60,12 @@
 
 removable=no
 efi_quiet=
+sb_var=/sys/firmware/efi/vars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c/data
+if [ -e "$sb_var" ] && [ "$(printf %x \'"$(cat "$sb_var")")" = 1 ]; then
+  uefi_secure_boot=yes
+else
+  uefi_secure_boot=no
+fi
 
 # Get GRUB_DISTRIBUTOR.
 if test -f "${sysconfdir}/default/grub" ; then
@@ -125,6 +131,12 @@
     cat <<EOF
    --removable             the installation device is removable
    --bootloader-id=ID      the ID of bootloader.
+   --uefi-secure-boot      install an image usable with UEFI Secure Boot
+                           (only available if the grub-efi-amd64-signed
+                           package is installed)
+   --no-uefi-secure-boot   do not install an image usable with UEFI Secure
+                           Boot, even if the system was currently started
+                           using it
 EOF
 fi
     cat <<EOF
@@ -250,6 +262,11 @@
     -f | --force)
         setup_force="--force" ;;
 
+    --uefi-secure-boot)
+	uefi_secure_boot=yes ;;
+    --no-uefi-secure-boot)
+	uefi_secure_boot=no ;;
+
     -*)
 	echo "Unrecognized option \`$option'" 1>&2
 	usage
@@ -352,39 +369,32 @@
         # not collide with other vendors.  To minimise collisions, we use the
         # name of our distributor if possible.
 	efi_distributor="$bootloader_id"
+	# It is convenient for each architecture to have a different
+	# efi_file, so that different versions can be installed in parallel.
+	case "$target_cpu" in
+	    i386)
+		efi_suffix=ia32 ;;
+	    x86_64)
+		efi_suffix=x64 ;;
+	# GRUB does not yet support these architectures, but they're defined
+	# by the specification so we include them here to ease future
+	# expansion.
+	    ia64)
+		efi_suffix=ia64 ;;
+	    *)
+		efi_suffix= ;;
+	esac
 	if test $removable = yes; then
       	   # The specification makes stricter requirements of removable
 	   # devices, in order that only one image can be automatically loaded
 	   # from them.  The image must always reside under /EFI/BOOT, and it
 	   # must have a specific file name depending on the architecture.
 	    efi_distributor=BOOT
-	    case "$target_cpu" in
-		i386)
-		    efi_file=BOOTIA32.EFI ;;
-		x86_64)
-		    efi_file=BOOTX64.EFI ;;
-	    # GRUB does not yet support these architectures, but they're defined
-	    # by the specification so we include them here to ease future
-	    # expansion.
-		ia64)
-		    efi_file=BOOTIA64.EFI ;;
-	    esac
+	    efi_file="BOOT$(printf %s "$efi_suffix" | tr a-z A-Z).EFI"
 	else
 	    # It is convenient for each architecture to have a different
 	    # efi_file, so that different versions can be installed in parallel.
-	    case "$target_cpu" in
-		i386)
-		    efi_file=grubia32.efi ;;
-		x86_64)
-		    efi_file=grubx64.efi ;;
-	 # GRUB does not yet support these architectures, but they're defined
- 	 # by the specification so we include them here to ease future
-	 # expansion.
-		ia64)
-		    efi_file=grubia64.efi ;;
-		*)
-		    efi_file=grub.efi ;;
-	    esac
+	    efi_file="grub$efi_suffix.efi"
 	   # TODO: We should also use efibootmgr, if available, to add a Boot
 	   # entry for ourselves.
 	fi
@@ -508,7 +518,7 @@
     # Strip partition number
     grub_partition="`echo "${grub_drive}" | sed -e 's/^[^,]*[,)]//; s/)$//'`"
     grub_drive="`echo "${grub_drive}" | sed -e s/,[a-z0-9,]*//g`"
-    if [ "$disk_module" = ata ] || [ "x${grub_drive}" != "x${install_drive}" ] || ([ "x$platform" != xefi ] && [ "x$platform" != xpc ] && [ x"${target_cpu}-${platform}" != x"sparc64-ieee1275" ]) ; then
+    if [ "$disk_module" = ata ] || [ "x${grub_drive}" != "x${install_drive}" ] || ([ "x$platform" != xefi ] && [ "x$platform" != xpc ] && [ x"${target_cpu}-${platform}" != x"sparc64-ieee1275" ]) || ([ "x$platform" = xefi ] && [ "$uefi_secure_boot" = yes ]); then
         # generic method (used on coreboot and ata mod)
         uuid="`"$grub_probe" --device-map="${device_map}" --target=fs_uuid --device "${grub_device}"`"
         if [ "x${uuid}" = "x" ] ; then
@@ -606,10 +616,28 @@
 	}
     fi
 elif [ x"$platform" = xefi ]; then
-    cp "${grubdir}/core.${imgext}" "${efidir}/${efi_file}"
-    # For old macs. Suggested by Peter Jones.
-    if [ x$target_cpu = xi386 ]; then
-	cp "${grubdir}/core.${imgext}" "${efidir}/boot.efi"
+    efi_signed="${pkglibdir}-signed/grub$efi_suffix.efi.signed"
+    if [ "$uefi_secure_boot" = yes ] && [ -e "$efi_signed" ]; then
+	shim_signed=/usr/lib/shim/shim.efi.signed
+	if [ -e "$shim_signed" ]; then
+	    if [ "$removable" != yes ]; then
+		efi_file="shim$efi_suffix.efi"
+	    fi
+	    cp "$shim_signed" "${efidir}/${efi_file}"
+	    cp "$efi_signed" "${efidir}/grub$efi_suffix.efi"
+	else
+	    cp "$efi_signed" "${efidir}/${efi_file}"
+	fi
+	if [ x"$config_opt_file" != x ]; then
+	    cp "${config_opt_file}" "${efidir}/grub.cfg"
+	    echo 'configfile $prefix/grub.cfg' >> "${efidir}/grub.cfg"
+	fi
+    else
+	cp "${grubdir}/core.${imgext}" "${efidir}/${efi_file}"
+	# For old macs. Suggested by Peter Jones.
+	if [ x$target_cpu = xi386 ]; then
+	    cp "${grubdir}/core.${imgext}" "${efidir}/boot.efi"
+	fi
     fi
 
     # Try to make this image bootable using the EFI Boot Manager, if available.
